File: desktop/macos/Sources/VoiceBridgeAI/Session/SessionController.swift
import AppKit

@MainActor
final class SessionController {
    static let shared = SessionController()

    let store = SubtitleStore()
    private(set) var isRunning = false
    private(set) var isStarting = false

    private let webSocket = WebSocketSession()
    private var capture: SystemAudioCapture?
    private var engineConfig = EngineConfig()
    private var silenceMonitor = PcmSilenceMonitor()

    private init() {
        store.onChange = { [weak self] in
            guard let self else { return }
            AppDelegate.shared?.lastSubtitleStore = self.store
            AppDelegate.shared?.overlay.update(with: self.store)
        }

        webSocket.onASR = { [weak self] payload in
            Task { @MainActor in
                self?.silenceMonitor.reset()
                self?.store.applyASR(payload)
                self?.recordTranslationIfNeeded(payload)
            }
        }
        webSocket.onError = { [weak self] message in
            Task { @MainActor in
                self?.store.showError(message)
                self?.stop()
            }
        }
    }

    func start() async -> String? {
        if isRunning || isStarting { return nil }

        isStarting = true
        AppDelegate.shared?.refreshControlUI()
        defer {
            isStarting = false
            AppDelegate.shared?.refreshControlUI()
        }

        let serverErr = await ServerManager.shared.ensureRunning()
        if let serverErr {
            return serverErr
        }

        do {
            try await SettingsStore.shared.refresh()
            engineConfig = SettingsStore.shared.engine
        } catch {
            return "无法读取引擎配置"
        }

        do {
            try await webSocket.connect(config: engineConfig)
        } catch {
            return error.localizedDescription
        }

        if #available(macOS 13.0, *) {
            if !ScreenCaptureAccess.ensureAccess() {
                webSocket.disconnect()
                ScreenCaptureAccess.openSystemSettings()
                return "需要屏幕录制权限才能采集系统音频。已尝试打开系统设置，请允许 VoiceBridgeAI 后重试。"
            }

            let cap = SystemAudioCapture()
            cap.onPCM = { [weak self] data in
                Task { @MainActor in
                    guard let self, self.isRunning else { return }
                    if self.silenceMonitor.feed(pcm: data), !self.store.segments.isEmpty {
                        self.store.clearDisplay()
                        self.silenceMonitor.reset()
                    }
                    do {
                        try await self.webSocket.send(pcm: data)
                    } catch {
                        self.store.showError("音频发送失败：\(error.localizedDescription)")
                        self.stop()
                    }
                }
            }
            cap.onFailure = { [weak self] message in
                Task { @MainActor in
                    self?.store.showError(message)
                    self?.stop()
                }
            }
            do {
                try await cap.start()
                capture = cap
            } catch {
                webSocket.disconnect()
                return "音频采集失败：\(error.localizedDescription)。请在 系统设置 → 隐私 → 屏幕录制 中允许本应用。"
            }
        } else {
            webSocket.disconnect()
            return "需要 macOS 13 或更高版本"
        }

        isRunning = true
        silenceMonitor.reset()
        store.reset()
        beginTranslationRecordingIfNeeded()
        AppDelegate.shared?.overlay.update(with: store)
        return nil
    }

    func stop() {
        if #available(macOS 13.0, *) {
            capture?.stop()
        }
        capture = nil
        webSocket.disconnect()
        isRunning = false
        isStarting = false
        TranslationRecorder.shared.endSession()
        store.hide()
        AppDelegate.shared?.overlay.update(with: store)
        AppDelegate.shared?.refreshControlUI()
    }

    func applyReviseMode(_ mode: String) async throws {
        SettingsStore.shared.engine.reviseMode = mode
        _ = try await SettingsStore.shared.saveEngine()
        try await reconfigureEngine()
        AppDelegate.shared?.overlay.update(with: store)
    }

    func reconfigureEngine() async throws {
        try await SettingsStore.shared.refresh()
        engineConfig = SettingsStore.shared.engine
        if isRunning {
            try await webSocket.reconfigure(config: engineConfig)
            AppDelegate.shared?.overlay.update(with: store)
        }
    }

    private func beginTranslationRecordingIfNeeded() {
        guard TranscriptPreferences.recordEnabled else { return }
        let health = SettingsStore.shared.health
        let modeId = engineConfig.reviseMode
        let label = ReviseModeGuides.info(for: modeId, health: health)?.label ?? modeId
        TranslationRecorder.shared.beginSession(reviseModeId: modeId, reviseModeLabel: label)
    }

    private func recordTranslationIfNeeded(_ payload: [String: Any]) {
        guard TranscriptPreferences.recordEnabled else { return }
        guard payload["final"] as? Bool == true else { return }
        let english = (payload["text"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !english.isEmpty else { return }
        let segmentId = String(describing: payload["segmentId"] ?? "")
        let chinese = (payload["translation"] as? String) ?? ""
        let revised = (payload["revise"] as? Bool ?? false) || (payload["lookback"] as? Bool ?? false)
        TranslationRecorder.shared.record(
            segmentId: segmentId,
            english: english,
            chinese: chinese,
            revised: revised
        )
    }
}

File: desktop/windows/VoiceBridgeAI/VoiceBridgeAI/Session/SessionController.cs
using VoiceBridgeAI.Capture;
using VoiceBridgeAI.Settings;

namespace VoiceBridgeAI.Session;

public sealed class SessionController
{
    public static SessionController Shared { get; } = new();

    public SubtitleStore Store { get; } = new();
    public bool IsRunning { get; private set; }
    public bool IsStarting { get; private set; }

    public event Action? StateChanged;

    private readonly WebSocketSession _webSocket = new();
    private SystemAudioCapture? _capture;
    private EngineConfig _engineConfig = new();
    private PcmSilenceMonitor _silenceMonitor = new();
    private readonly SynchronizationContext? _uiContext;

    private SessionController()
    {
        _uiContext = SynchronizationContext.Current;

        Store.Changed += () => PostToUi(() => StateChanged?.Invoke());

        _webSocket.AsrReceived += payload => PostToUi(() =>
        {
            _silenceMonitor.Reset();
            Store.ApplyAsr(payload);
            RecordTranslationIfNeeded(payload);
        });

        _webSocket.ErrorReceived += message => PostToUi(() =>
        {
            Store.ShowError(message);
            Stop();
        });
    }

    public async Task<string?> StartAsync()
    {
        if (IsRunning || IsStarting)
        {
            return null;
        }

        IsStarting = true;
        NotifyStateChanged();
        try
        {
            var serverErr = await ServerManager.Shared.EnsureRunningAsync();
            if (serverErr is not null)
            {
                return serverErr;
            }

            try
            {
                await SettingsStore.Shared.RefreshAsync();
                _engineConfig = SettingsStore.Shared.Engine;
            }
            catch
            {
                return "无法读取引擎配置";
            }

            try
            {
                await _webSocket.ConnectAsync(_engineConfig);
            }
            catch (Exception ex)
            {
                return ex.Message;
            }

            var capture = new SystemAudioCapture();
            capture.PcmAvailable += async pcm =>
            {
                if (!IsRunning)
                {
                    return;
                }

                if (_silenceMonitor.Feed(pcm) && Store.Segments.Count > 0)
                {
                    Store.ClearDisplay();
                    _silenceMonitor.Reset();
                }

                try
                {
                    await _webSocket.SendPcmAsync(pcm);
                }
                catch (Exception ex)
                {
                    PostToUi(() =>
                    {
                        Store.ShowError($"音频发送失败：{ex.Message}");
                        Stop();
                    });
                }
            };

            capture.Failed += message => PostToUi(() =>
            {
                Store.ShowError(message);
                Stop();
            });

            try
            {
                capture.Start();
            }
            catch (Exception ex)
            {
                await _webSocket.DisconnectAsync();
                return $"音频采集失败：{ex.Message}。请确认系统正在播放音频，且未被其他录音软件独占。";
            }

            _capture = capture;
            IsRunning = true;
            _silenceMonitor.Reset();
            Store.Reset();
            BeginTranslationRecordingIfNeeded();
            return null;
        }
        finally
        {
            IsStarting = false;
            NotifyStateChanged();
        }
    }

    public void Stop()
    {
        _capture?.Stop();
        _capture?.Dispose();
        _capture = null;
        _ = _webSocket.DisconnectAsync();
        IsRunning = false;
        IsStarting = false;
        TranslationRecorder.Shared.EndSession();
        Store.Hide();
        NotifyStateChanged();
    }

    private void NotifyStateChanged()
    {
        PostToUi(() => StateChanged?.Invoke());
    }

    private void PostToUi(Action action)
    {
        if (_uiContext is null)
        {
            action();
            return;
        }

        _uiContext.Post(_ => action(), null);
    }

    private void BeginTranslationRecordingIfNeeded()
    {
        if (!TranscriptPreferences.RecordEnabled)
        {
            return;
        }

        var modeId = _engineConfig.ReviseMode;
        var label = ReviseModeLabel(modeId);
        TranslationRecorder.Shared.BeginSession(modeId, label);
    }

    private static void RecordTranslationIfNeeded(IReadOnlyDictionary<string, object?> payload)
    {
        if (!TranscriptPreferences.RecordEnabled)
        {
            return;
        }

        if (payload.TryGetValue("final", out var finalVal) && finalVal is not true)
        {
            return;
        }

        var english = payload.TryGetValue("text", out var textVal)
            ? Convert.ToString(textVal)?.Trim() ?? ""
            : "";
        if (string.IsNullOrEmpty(english))
        {
            return;
        }

        var segmentId = payload.TryGetValue("segmentId", out var segVal)
            ? Convert.ToString(segVal) ?? ""
            : "";
        var chinese = payload.TryGetValue("translation", out var trVal)
            ? Convert.ToString(trVal) ?? ""
            : "";
        var revised = (payload.TryGetValue("revise", out var reviseVal) && reviseVal is true)
            || (payload.TryGetValue("lookback", out var lookbackVal) && lookbackVal is true);

        TranslationRecorder.Shared.Record(segmentId, english, chinese, revised);
    }

    private static string ReviseModeLabel(string modeId)
    {
        try
        {
            var health = SettingsStore.Shared.Health;
            return ProviderOption.List(health, "reviseModes")
                .FirstOrDefault(p => p.Id == modeId).Label ?? modeId;
        }
        catch
        {
            return modeId;
        }
    }
}
